### PR TITLE
std: Fix partial writes in LineWriter

### LINKED ISSUE #37807: BrokenPipe sometimes not reported
Try piping this program's stdout to `head`. After the first 10 lines, `head` will exit. The pipe is now broken, so the next write should fail, and we should get `thread 'main' panicked at 'failed printing to stdout: Broken pipe (os error 32)'`

Instead, the Rust program keeps running indefinitely. GDB says `SIGPIPE` is "received" (but I guess ignored) by the process every time it tries to write.

Feels like a race condition: if you remove the `sleep` or just change it to be short, `println!` panics as expected.

```rust
// usage: ./yep | head
use std::time::Duration;
use std::thread::sleep;

fn main() {
    loop {
        println!("y");  // should panic after `head` exits
        sleep(Duration::new(0, 250_000_000));
    }
}
```

tested on Linux, `rustc 1.15.0-nightly (0ed951993 2016-11-14)`

### PR DESCRIPTION
**Pull request by alexcrichton**
Previously the `LineWriter` could successfully write some bytes but then fail to
report that it has done so. Additionally, an erroneous flush after a successful
write was permanently ignored. This commit fixes these two issues by (a)
maintaining a `need_flush` flag to indicate whether a flush should be the first
operation in `LineWriter::write` and (b) avoiding returning an error once some
bytes have been successfully written.

Closes #37807

### CODE REVIEW

**Comment by Stebalien** [comment]
> Previously the `LineWriter` could successfully write some bytes but then fail to
report that it has done so. 

Previously as in before #32107 (or am I missing something)?

**Comment by Stebalien** [comment]
Really, we should probably avoid flushing unless the last character written was a newline (that's kind of the point of this wrapper). Unfortunately, that could lead to a pathological error swallowing case if we keep making small, partial writes that never finish on a newline.

**Comment by alexcrichton** [comment]
@Stebalien oh I meant previously as in just before this commit. I think this was likely accidental but if [this write fails](https://github.com/rust-lang/rust/blob/f50dbd580f3d5d88a300da540db3c85164bfd0de/src/libstd/io/buffered.rs#L777) then we didn't report the successfully written bytes [earlier](https://github.com/rust-lang/rust/blob/f50dbd580f3d5d88a300da540db3c85164bfd0de/src/libstd/io/buffered.rs#L772).

> Really, we should probably avoid flushing unless the last character written was a newline

Isn't that essentially what happens here? After a successful write of data which may contain a newline we guarantee a flush afterwards?

**Comment by alexcrichton** [comment]
Er well so there's two conditions after the partial write:

* Did we write all the data?
* Did the flush succeed?

If the former is false then we *may* have written a newline (we don't know unless we search). Do you think in the case of a partial write we should search for a newline to see if we should actually flush? It seems like a rare enough case that somewhat eager flushing would also be fine...

**Comment by Stebalien** [comment]
> @Stebalien oh I meant previously as in just before this commit. I think this was likely accidental but if this write fails then we didn't report the successfully written bytes earlier.

head -> desk: I completely missed that.

> If the former is false then we may have written a newline (we don't know unless we search). Do you think in the case of a partial write we should search for a newline to see if we should actually flush? It seems like a rare enough case that somewhat eager flushing would also be fine...

After thinking about it, your way is probably best. The only way to get a partial write from a `BufWriter` is if we try to write more than the inner buffer can hold in one go (so, we won't end up flushing for small writes). In that case, the `BufWriter` will write directly to the wrapped writer (bypassing the buffer). A subsequent flush won't flush anything out of the internal buffer (it will have been flushed before the direct write) and instead just call flush on the wrapped writer. In the normal case where the wrapped writer is some kind of os stream (e.g. a file or socket), the flush will be a no-op.

TL;DR: 

1. The extra flush will only happen for very large writes.
2. It will usually be a no-op (the buffer will already have been automatically flushed by the `BufWriter`).

**Comment by alexcrichton** [comment]
Makes sense to me!

**Comment by brson** [comment]
@bors r+

It's a clever patch. Nice work.
